analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.RoslynCfg.cs

```diff
@@ -0,0 +1,151 @@
+                        walker.CheckPaths();
+                    }
+                }
+            }
+
+            public void CheckForNoExitMethod(SyntaxNodeAnalysisContext c, CSharpSyntaxNode body, SyntaxToken identifier, ISymbol symbol)
+            {
+                var operation = c.SemanticModel.GetOperation(body.Parent);
+                var cfg = ControlFlowGraph.Create(operation.RootOperation().Syntax, c.SemanticModel);
+                var enclosingKinds = new HashSet<SyntaxKind>
+                {
+                    SyntaxKindEx.LocalFunctionStatement, SyntaxKind.SimpleLambdaExpression, SyntaxKind.AnonymousMethodExpression, SyntaxKind.ParenthesizedLambdaExpression
+                };
+                if (c.Node.IsKind(SyntaxKindEx.LocalFunctionStatement))
+                {
+                    // we need to go up and track all possible enclosing local function statements
+                    foreach (var enclosingFunction in c.Node.Ancestors().Where(x => enclosingKinds.Contains(x.Kind())).Reverse())
+                    {
+                        if (enclosingFunction.IsKind(SyntaxKindEx.LocalFunctionStatement))
+                        {
+                            cfg = cfg.GetLocalFunctionControlFlowGraph(c.SemanticModel.GetDeclaredSymbol(enclosingFunction) as IMethodSymbol);
+                        }
+                        else
+                        {
+                            var operationWrapper = cfg.FlowAnonymousFunctionOperations().Single(x => x.WrappedOperation.Syntax == enclosingFunction);
+                            cfg = cfg.GetAnonymousFunctionControlFlowGraph(operationWrapper);
+                        }
+                        if (cfg == null)
+                        {
+                            return;
+                        }
+                    }
+
+                    cfg = cfg.GetLocalFunctionControlFlowGraph(symbol as IMethodSymbol);
+                    if (cfg == null)
+                    {
+                        return;
+                    }
+                }
+
+                var context = new RecursionContext<ControlFlowGraph>(cfg, symbol, identifier.GetLocation(), c, "method's recursion");
+                var walker = new RecursionSearcher(context);
+                walker.CheckPaths();
+            }
+
+            private class RecursionSearcher : CfgAllPathValidator
+            {
+                private readonly RecursionContext<ControlFlowGraph> context;
+                private readonly bool isGetAccesor;
+
+                public RecursionSearcher(RecursionContext<ControlFlowGraph> context, bool isGetAccesor = true)
+                    : base(context.ControlFlowGraph)
+                {
+                    this.context = context;
+                    this.isGetAccesor = isGetAccesor;
+                }
+
+                public void CheckPaths()
+                {
+                    if (CheckAllPaths() || CfgHasUnescapableLoop())
```

Comment:
Nice simplification!

_Minor performance:_ The loop part seems cheaper than diving deep into operations. But it will not help in too many cases.
```suggestion
                    if (CfgHasUnescapableLoop() || CheckAllPaths())
```